Address image compatibility checks in Scala API

pbylicki opened the pull request:
Based on discussion in #139 I added changes to `PostgreSQLContainer` allowing initializing it with `DockerImageName` instead of `String`. If you are OK with that, I can extend this PR with the same changes applied to remaining modules.

1. Changed main constructor to use `DockerImageName`
2. Added auxiliary constructor using String as image name for backward compatibility with deprecated annotation. Removed default value to prevent ambiguity when initializing with default arguments only.
3. Changed default `apply` method to use `DockerImageName`
4. Added auxiliary `apply` method using String as image name for backward compatibility without deprecated annotation. Removed default value to prevent ambiguity when initializing with default arguments only.
5. Changed `Def` `dockerImageName` field to `DockerImageName`
6. Added auxiliary `Def.apply` method using String as image name for backward compatibility without deprecated annotation. Removed default value to prevent ambiguity when initializing with default arguments only.

@dimafeng @LMnet  Let me know if anything is missing or excessive here.

dimafeng commented on modules/postgres/src/main/scala/com/dimafeng/testcontainers/PostgreSQLContainer.scala:
```diff
@@ -91,4 +128,25 @@ object PostgreSQLContainer {
   }
+
+  object Def {
+    def apply(
```
This one should be deprecated too, right?

dimafeng commented:
@pbylicki looks good to me! 👍

pbylicki commented:
@dimafeng I missed one important detail before. With the approach described above we end up with multiple overloaded alternatives of method defined with default arguments which is forbidden by scala compiler.

One solution that I see is to:
1. Change main constructor to use `DockerImageName` and remove ALL default arguments
2. Turn old constructor into auxiliary constructor using String as image name with default arguments for backward compatibility. I am not sure if I should add deprecated annotation if it stays as the only constructor with default arguments.
3. Change `Def` `dockerImageName` field to `DockerImageName` without maintaining backward compatibility as new API is still labeled as experimental

If that's not enough some new custom constructor methods would need to be introduced.

Another issue is that non-standard Container implementations use different names for docker image name argument and some of them don't expect full image name (repo + tag), but only tag (examples are InfluxDBContainer and PulsarContainer). I thought about unifying this argument name to `dockerImageName` in main constructor and creating getter methods with deprecated annotation for those Containers that are case classes and expose them.
I pushed changes to mentioned Container classes to demonstrate.

dimafeng commented:
Maybe not the best option but what if we use an implicit conversion from String to DockerImageName. In this case, we can change all constructors to use DockerImageName.

@LMnet what do you think?

LMnet commented:
I think that we definitely should retain default arguments in the constructors. And at the same time, I think we should not break backward compatibility.

I don't think that implicit conversion is a good decision, but comparing to others it's the best solution. Moreover, it looks like `DockerImageName.parse` function is not throwing exceptions and can consume any string (even empty string, which is definitely a bug). So, it will be a pretty safe approach.

In future versions, we may deprecate this implicit conversion and force users to rewrite their constructor.

But we should be careful with an implicit scope. Users should not export any orphaned implicits to make the new implicit conversion work.

pbylicki commented:
@dimafeng @LMnet  Thanks for suggestions, changed all constructors that used full image name before and added implicit conversion method.
What about Container classes that use only version as argument? Should we leave them unchanged since existing API is not affected by image compatibility checks?

dimafeng commented:
@pbylicki I'd leave them as-is for now (as long as it matches the reflective Java constructors)

dimafeng commented on core/src/main/scala/com/dimafeng/testcontainers/implicits/DockerImageNameConverters.scala:
```diff
@@ -0,0 +1,10 @@
+import org.testcontainers.utility.DockerImageName
+
+trait DockerImageNameConverters {
+  implicit def stringToDockerImageName(s: String): DockerImageName =
```
Should we annotate this conversion as deprecated?

pbylicki replied:
Sure, will add it

dimafeng commented on modules/oracle/src/test/scala/com/dimafeng/testcontainers/OracleSpec.scala:
```diff
@@ -2,6 +2,7 @@ package com.dimafeng.testcontainers
 
 import java.sql.DriverManager
 
+import com.dimafeng.testcontainers.implicits._
```
(I haven't used scala for a while so my memory on implicit resolution is not that fresh) Can we somehow workaround this import so the implicit conversion comes without explicit import (e.g. define the conversion method in the base class)?

pbylicki replied:
For custom type we could solve it by adding implicit conversion to companion object, but we have plain String here.
One solution would be to add `DockerImageNameConverters` trait to `TestContainersSuite` in scalatest and munit modules. That should allow switching to new version without any code changes in most cases. Is that acceptable?

dimafeng replied:
That should work. @LMnet do you have thoughts on this?

LMnet replied:
That's exactly what I mean by saying: "Users should not export any orphaned implicits to make the new implicit conversion work."

I thought about adding this implicit conversion to all container companions, but the solution with `TestContainersSuite` is much better. Go for it @pbylicki

pbylicki replied:
Thanks, added.

pbylicki commented:
@dimafeng Added implicit conversions to ScalaTest and MUnit traits, updated documentation and removed draft from PR.

LMnet commented:
lgtm

dimafeng commented:
@pbylicki @LMnet thank you!
